### ISSUE TITLE
mailman-core terminates with error and stops instead of waiting and retrying when MySQL database is not available (yet)

### ISSUE THREAD

**Issue by JonasT** [maxking/docker-mailman#138]
mailman-core terminates with error and stops instead of waiting and retrying when database is not available (yet). This is problematic in a docker environment because the database container might still be initializing and this could lead to unexpected startup failures.

Of course there's always the `restart: always` workaround, but it would be a lot nicer if the container handled a database connection failure properly with a sleep-a-few-seconds-and-retry loop instead of what is currently happening:

```
mailman-core_1    | Traceback (most recent call last):
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool.py", line 1122, in _do_get
mailman-core_1    |     return self._pool.get(wait, self._timeout)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/queue.py", line 145, in get
mailman-core_1    |     raise Empty
mailman-core_1    | sqlalchemy.util.queue.Empty
mailman-core_1    | 
mailman-core_1    | During handling of the above exception, another exception occurred:
mailman-core_1    | 
mailman-core_1    | Traceback (most recent call last):
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 2147, in _wrap_pool_connect
mailman-core_1    |     return fn()
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool.py", line 387, in connect
mailman-core_1    |     return _ConnectionFairy._checkout(self)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool.py", line 766, in _checkout
mailman-core_1    |     fairy = _ConnectionRecord.checkout(pool)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool.py", line 516, in checkout
mailman-core_1    |     rec = pool._do_get()
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool.py", line 1138, in _do_get
mailman-core_1    |     self._dec_overflow()
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/langhelpers.py", line 66, in __exit__
mailman-core_1    |     compat.reraise(exc_type, exc_value, exc_tb)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 187, in reraise
mailman-core_1    |     raise value
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool.py", line 1135, in _do_get
mailman-core_1    |     return self._create_connection()
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool.py", line 333, in _create_connection
mailman-core_1    |     return _ConnectionRecord(self)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool.py", line 461, in __init__
mailman-core_1    |     self.__connect(first_connect_check=True)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool.py", line 651, in __connect
mailman-core_1    |     connection = pool._invoke_creator(self)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/strategies.py", line 105, in connect
mailman-core_1    |     return dialect.connect(*cargs, **cparams)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 393, in connect
mailman-core_1    |     return self.dbapi.connect(*cargs, **cparams)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/MySQLdb/__init__.py", line 86, in Connect
mailman-core_1    |     return Connection(*args, **kwargs)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/MySQLdb/connections.py", line 204, in __init__
mailman-core_1    |     super(Connection, self).__init__(*args, **kwargs2)
mailman-core_1    | _mysql_exceptions.OperationalError: (2003, 'Can\'t connect to MySQL server on \'mailman-db\' (111 "Connection refused")')
mailman-core_1    | 
mailman-core_1    | The above exception was the direct cause of the following exception:
mailman-core_1    | 
mailman-core_1    | Traceback (most recent call last):
mailman-core_1    |   File "/usr/local/bin/mailman", line 11, in <module>
mailman-core_1    |     sys.exit(main())
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/mailman/bin/mailman.py", line 97, in main
mailman-core_1    |     initialize(config_path)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/mailman/core/initialize.py", line 185, in initialize
mailman-core_1    |     initialize_2(propagate_logs=propagate_logs)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/mailman/core/initialize.py", line 156, in initialize_2
mailman-core_1    |     config.db = getUtility(IDatabaseFactory, utility_name).create()
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/mailman/database/factory.py", line 55, in create
mailman-core_1    |     SchemaManager(database).setup_database()
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/mailman/database/factory.py", line 83, in setup_database
mailman-core_1    |     context = MigrationContext.configure(self._database.store.connection())
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 998, in connection
mailman-core_1    |     execution_options=execution_options)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 1003, in _connection_for_bind
mailman-core_1    |     engine, execution_options)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 403, in _connection_for_bind
mailman-core_1    |     conn = bind.contextual_connect()
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 2112, in contextual_connect
mailman-core_1    |     self._wrap_pool_connect(self.pool.connect, None),
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 2151, in _wrap_pool_connect
mailman-core_1    |     e, dialect, self)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1465, in _handle_dbapi_exception_noconnection
mailman-core_1    |     exc_info
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause
mailman-core_1    |     reraise(type(exception), exception, tb=exc_tb, cause=cause)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 186, in reraise
mailman-core_1    |     raise value.with_traceback(tb)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 2147, in _wrap_pool_connect
mailman-core_1    |     return fn()
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool.py", line 387, in connect
mailman-core_1    |     return _ConnectionFairy._checkout(self)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool.py", line 766, in _checkout
mailman-core_1    |     fairy = _ConnectionRecord.checkout(pool)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool.py", line 516, in checkout
mailman-core_1    |     rec = pool._do_get()
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool.py", line 1138, in _do_get
mailman-core_1    |     self._dec_overflow()
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/langhelpers.py", line 66, in __exit__
mailman-core_1    |     compat.reraise(exc_type, exc_value, exc_tb)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 187, in reraise
mailman-core_1    |     raise value
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool.py", line 1135, in _do_get
mailman-core_1    |     return self._create_connection()
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool.py", line 333, in _create_connection
mailman-core_1    |     return _ConnectionRecord(self)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool.py", line 461, in __init__
mailman-core_1    |     self.__connect(first_connect_check=True)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool.py", line 651, in __connect
mailman-core_1    |     connection = pool._invoke_creator(self)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/strategies.py", line 105, in connect
mailman-core_1    |     return dialect.connect(*cargs, **cparams)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 393, in connect
mailman-core_1    |     return self.dbapi.connect(*cargs, **cparams)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/MySQLdb/__init__.py", line 86, in Connect
mailman-core_1    |     return Connection(*args, **kwargs)
mailman-core_1    |   File "/usr/local/lib/python3.6/site-packages/MySQLdb/connections.py", line 204, in __init__
mailman-core_1    |     super(Connection, self).__init__(*args, **kwargs2)
mailman-core_1    | sqlalchemy.exc.OperationalError: (_mysql_exceptions.OperationalError) (2003, 'Can\'t connect to MySQL server on \'mailman-db\' (111 "Connection refused")')
```
(followed by a container shutdown instead of reasonable wait time + retry)

**Comment by maxking** [comment]
There is some bash magic for Postgres to do this in both the images. #125 is already proposed to add this for MySQL as well.

It would be interesting to find something that would do it in database-type independent way so that we don't have to do it separately for MySQL and Postgres.

**Comment by maxking** [comment]
Fixed in #140!
